File: redbird_recon/s4_sweep.m
% S4_SWEEP  Using the cached two-surface operator, sweep depth weighting (beta) and
% Tikhonov (alpha) to suppress scalp leakage and localize the cortex. No forward.
here = fileparts(mfilename('fullpath'));
addpath('/home/users/fangq/space/git/Project/github/redbird-m/matlab');
addpath('/home/users/fangq/space/git/Project/jsonlab');

S = load(fullfile(here, 's4_op.mat'));               % Anorm, cmeas, ncortex, nscalp
Anorm = S.Anorm; cmeas = S.cmeas; ncortex = S.ncortex; nscalp = S.nscalp; Nsurf = ncortex + nscalp;
m = loadbj(fullfile(here, 'hrf_measurement.jdb'));
wl = double(m.wavelengths(:))'; nwl = numel(wl);
dOD = double(m.dOD); pairs = double(m.pairs); npair = size(pairs, 1);
taxis = double(m.time(:)); [~, tidx] = min(abs(taxis - 10.0));
gt = loadbj(fullfile(here, 'groundtruth.bgii'));
cortex = double(gt.GIFTIData.MeshVertex3.Data);
trC3 = double(gt.GIFTIData.MeshVertex3.Properties.HbO_C3);
trC4 = double(gt.GIFTIData.MeshVertex3.Properties.HbO_C4);
itC3 = find(trC3 == max(trC3), 1); itC4 = find(trC4 == max(trC4), 1);

wrow = 1 ./ sqrt(cmeas + 1e-6 * max(cmeas));
rhsC3 = zeros(npair * nwl, 1); rhsC4 = rhsC3;
for w = 1:nwl
    rhsC3((w-1)*npair + (1:npair)) = -dOD(:, w, tidx, 1);
    rhsC4((w-1)*npair + (1:npair)) = -dOD(:, w, tidx, 2);
end

fprintf('%6s %6s | %-7s %-7s %-6s | %-7s %-7s %-6s\n', 'alpha','beta','errC3','corrC3','leakC3','errC4','corrC4','leakC4');
for alpha = [0.003 0.01 0.05]
  for beta = [1 0.1 0.03 0.01 0.003]
    Aw = (wrow .* Anorm);
    cn = sqrt(sum(Aw.^2, 1))'; Wc = 1 ./ (cn + beta * max(cn));
    Aw = Aw .* Wc';
    lam = alpha * max(sum(Aw.^2, 2));
    out = zeros(1, 6);
    rhss = {rhsC3, rhsC4}; its = [itC3 itC4]; trs = {trC3, trC4};
    for t = 1:2
        z = rbreginv(Aw, wrow .* rhss{t}, lam); d = z .* Wc; h = d(1:Nsurf);
        cv = h(1:ncortex); sv = h(ncortex+(1:nscalp));
        [~, ir] = max(cv); cc = corrcoef(cv, trs{t});
        out((t-1)*3 + (1:3)) = [norm(cortex(ir,:) - cortex(its(t),:)), cc(1,2), max(abs(sv))/max(abs(cv))];
    end
    fprintf('%6.3g %6.3g | %7.1f %7.3f %6.2f | %7.1f %7.3f %6.2f\n', alpha, beta, out(1),out(2),out(3),out(4),out(5),out(6));
  end
end
exit

File: redbirdpy_recon/s4_sweep.py
#!/usr/bin/env python3
"""s4_sweep.py -- redbirdpy port of redbird_recon/s4_sweep.m.

Using the cached two-surface operator (s4_op.jdb, built by s4_recon_surface.py),
sweep depth weighting (beta) and Tikhonov (alpha) to trade off scalp-leakage
suppression against cortical localization. No forward solve is run.

Usage:  python3 s4_sweep.py
"""

import jdata as jd
import numpy as np
import redbirdpy as rb

import nirs_common as nc


def main():
    cachef = nc.HERE / "s4_op.jdb"
    if not cachef.exists():
        raise SystemExit("missing %s; run s4_recon_surface.py first" % cachef.name)
    S = jd.loadjd(str(cachef))
    Anorm, cmeas = np.asarray(S["Anorm"]), np.asarray(S["cmeas"])
    ncortex, nscalp = int(S["ncortex"]), int(S["nscalp"])
    Nsurf = ncortex + nscalp

    meas = nc.load_meas()
    cortex, _, props, _ = nc.load_truth()
    pairs, wlkey, dOD, tidx = meas["pairs"], meas["wlkey"], meas["dOD"], meas["tidx"]
    npair, nwl = pairs.shape[0], len(wlkey)
    trC3, trC4 = props["HbO_C3"], props["HbO_C4"]
    itC3, itC4 = np.argmax(trC3), np.argmax(trC4)

    wrow = 1.0 / np.sqrt(cmeas + 1e-6 * cmeas.max())
    rhsC3 = np.zeros(npair * nwl)
    rhsC4 = np.zeros(npair * nwl)
    for w in range(nwl):
        rhsC3[w * npair : (w + 1) * npair] = -dOD[:, w, tidx, 0]
        rhsC4[w * npair : (w + 1) * npair] = -dOD[:, w, tidx, 1]

    print(
        "%6s %6s | %-7s %-7s %-6s | %-7s %-7s %-6s"
        % ("alpha", "beta", "errC3", "corrC3", "leakC3", "errC4", "corrC4", "leakC4")
    )
    for alpha in (0.003, 0.01, 0.05):
        for beta in (1, 0.1, 0.03, 0.01, 0.003):
            Aw = wrow[:, None] * Anorm
            cn = np.sqrt(np.sum(Aw**2, axis=0))
            Wc = 1.0 / (cn + beta * cn.max())
            Aw = Aw * Wc[None, :]
            lam = alpha * np.max(np.sum(Aw**2, axis=1))
            out = []
            for rhs, it, tr in ((rhsC3, itC3, trC3), (rhsC4, itC4, trC4)):
                d = rb.reginv(Aw, wrow * rhs, lam) * Wc
                h = d[:Nsurf]
                cv, sv = h[:ncortex], h[ncortex:]
                ir = np.argmax(cv)
                cc = np.corrcoef(cv, tr)[0, 1]
                out += [
                    np.linalg.norm(cortex[ir] - cortex[it]),
                    cc,
                    np.max(np.abs(sv)) / np.max(np.abs(cv)),
                ]
            print(
                "%6.3g %6.3g | %7.1f %7.3f %6.2f | %7.1f %7.3f %6.2f"
                % (alpha, beta, out[0], out[1], out[2], out[3], out[4], out[5])
            )


if __name__ == "__main__":
    main()
